Add support to TCP and UDP protocols

ThibaudCartegnie opened the pull request:
# :sparkles:  Add support to TCP and UDP protocols #224
:bangbang: Once all the **checklist** is **done** you have to:
  * **stash merge** this pull request
  * **delete** the corresponding **branch**
  * **close** the associated **issue**

## :page_with_curl: Type of change

**Breaking change**: fix or feature that would cause existing functionality to not work as expected.

## :black_nib: Description

This PR adds the support for udp and tcp besides Unix sockets.

There are 2 major changes for the protocol :
 - The Darwin Header no longer contains a Certitude by default and it is now defined as packed, lowering its size from 56 bytes to 48 bytes.
 - The Manager configuration file holds a 'network' field to specify how the filter receives data (Unix Socket, TCP or UDP)

This draft is pushed for early comments.

## :dart: Test Environments

### Ubuntu (20.04)
- Redis (5.0.7)
- Boost (1.71.0)
- g++ (9.3)
- CMake (3.16.3)
- Python (3.8.5)
- Valgrind (3.15.0)


### HardenedBSD (12.2)
- Redis (6.0.14)
- Boost (1.72.0)
- clang (10.0.1)
- CMake (3.20.4)
- Python (3.8.10)
- Valgrind(N/A)

## :heavy_check_mark: Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] (**If new filter**) I have added corresponding page to the documentation
- [x] (**If other changes**) I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

</br>

- [x] :raising_hand: **I certify on my honor that all the information provided is true, and I've done all I can to deliver a high quality code**

Commit:
separated unix socket logic from generic logic in the server class into 2 classes

ThibaudCartegnie committed:
Separated logic between tasks and Sessions, the separation is only applied on the dga filter for now

ThibaudCartegnie committed:
Fixed potential corruption :
If more than 1 certitude is sent back to client AND a body is set, the body is copied on the memory location of the certitudes
It is fixed by using the correct offset when copying the body data.
The other function using similar logic has been modified to keep it coherent.

ThibaudCartegnie committed:
Separated Sessions into UnixSession and TcpSession
UnixSessions are spawn by UnixServer and TcpSessions by TcpServer
Included a modification when passing arguments from sessions to task (move instead of ref)

ThibaudCartegnie committed:
WIP : First draft of DarwinPacket struct
 - Used to pass needed data from Session to Task

ThibaudCartegnie committed:
Finished integration of DarwinPacket in session
(some work is still need for the tasks)

ThibaudCartegnie committed:
Modified DarwinPacket to be more OOP-like
State accessible by const getters only
State mutable with only two methods : add Certitude and mutateBody

ThibaudCartegnie committed:
Re-wired some things in the tasks
Added a few methods for DarwinPacket

ThibaudCartegnie committed:
Put back structs in c header
Added ParseBody in ATask::run with error handling

ThibaudCartegnie committed:
Adapted all compiled filters
Fixed issues with threadpool instance and size issues in the ASession

ThibaudCartegnie committed:
Added parsing of tcp/udp
Fixed issues in the serialization
Added support for ipv6 in Tcp sessions

ThibaudCartegnie committed:
Adapted TcpServer and Config for ip addresses
Adapted python manager to handle tcp sockets

ThibaudCartegnie committed:
DarwinPacket: fixed issues with the serialization
The packet was illformed near the certitudes
NextFilterConnector: Working WIP

ThibaudCartegnie committed:
Separated NextFilterConnector in :
Abstract/Tcp/Unix, fixed configuration for nextFilter

ThibaudCartegnie committed:
Moved all network related objects to network folder

ThibaudCartegnie committed:
Fixed Set of asio buffer in ANextFilterConnector

ThibaudCartegnie committed:
Fixed memory issues:
- DarwinPacket copied only the first 15 bytes of event id
- Destructor of AServer wasn't virtual leading to memory leak when
UnixServer was destroyed

ThibaudCartegnie committed:
Fixed problem with the manager
Service.update did not update nettwork field

ThibaudCartegnie committed:
Fixed manager (failed previous commit)

ThibaudCartegnie committed:
RedisManager: modified Rate Limit connection logic
Added retry attempts, modified tests accordingly

ThibaudCartegnie committed:
Adapted ATask, Asession and DarwinPacket
Proprietary filters needed additional methods for accessing private
fields of a packet

ThibaudCartegnie committed:
Adapted tests to TCP :
Added possibility to run all tests in TCP mode

ThibaudCartegnie committed:
Added tests for TCP connections
Assured working with ipv4 and ipv6 addresses
Fixed a few log lines in the redis test file

ThibaudCartegnie committed:
Added UDP protocol to darwin
Fixed a few problems and cleaned some code
Added tests for tcp and udp
Added possibility to run all test in unix or tcp

ThibaudCartegnie committed:
fixed few issues with the manager

ThibaudCartegnie committed:
Added comments documentation
Removed unused methods

ThibaudCartegnie committed:
Fixed method names

ThibaudCartegnie committed:
Few modification of NextFilter handling
Specification of move semantic for DarwinPacket

ThibaudCartegnie committed:
Few fixes when passing information to next filters

ThibaudCartegnie committed:
Added a tests for multi filters using the manager
Fixed a fex things accordingly

ThibaudCartegnie committed:
Fixed issue with UUIDs
EvtIdToString passed from Session to DarwinPacket

ThibaudCartegnie committed:
Fixed problems with next filters, cleaned tests
Also cleaned ASession a bit

ThibaudCartegnie committed:
Modified config of next filter for easier configuration

ThibaudCartegnie committed:
Modified protocol.h
added packed attribute to darwin header
Removed certitude list default size
Modified parsing and serializing accordingly

ThibaudCartegnie committed:
Fixed clang/gcc warnings
Most case were unused variables or poorly chosen types

ThibaudCartegnie committed:
Fixed compilation issues on hardenedBSD
Fixed issue with positional flags between test/manager and filters

ThibaudCartegnie committed:
Fixed last issue with UDP NextFilterConnector

ThibaudCartegnie committed:
Added Thread Pool Cpp dependency

ThibaudCartegnie commented on CMakeLists.txt:
```diff
@@ -72,12 +73,15 @@ set(
 	${Boost_INCLUDE_DIRS}
 )
 
+add_compile_options(-Wno-unknown-pragmas)
```
The pragmas are inside thrid party library, we can specify this flag for the specific library if it is necessary, should I open another PR/issue for this change?

ThibaudCartegnie commented on toolkit/Network.cpp:
```diff
@@ -91,5 +93,53 @@ namespace darwin {
+                return false;
+            }
+
+            if (port < 0 || port > 65353) {
+                DARWIN_LOG_ERROR("Network::ParsePort:: Error while parsing the port number : out of bounds [0; 65353]: '" + str_port + "'");
```
It is a mistake or a typo

ThibaudCartegnie commented on samples/protocol.h:
```diff
@@ -54,8 +53,9 @@ typedef struct {
     size_t                              certitude_size; //!< The size of the list containing the certitudes.
-    unsigned int                        certitude_list[DEFAULT_CERTITUDE_LIST_SIZE]; //!< The scores or the certitudes of the module. May be used to pass other info in specific cases.
-} darwin_filter_packet_t;
+    unsigned int                        certitude_list[0]; //!< DEPRECATED, certitudes are no longer passed through this array, 
```
This changed was made with @HugoSoszynski, it fixes a bug with the old way of handling this (there was an invalid read if  there is a body and multiple certitudes) 
But yes, it is a breaking change.

frikilax replied:
alright, thanks for the clarification

ThibaudCartegnie commented on samples/ftest/TestTask.cpp:
```diff
@@ -51,29 +52,37 @@ void TestTask::operator()() {
                 }
             }
-            else {
+            else if (_line == "load_test") {
+                DARWIN_LOG_DEBUG("TestTask:: load_test triggered, doing work for some times (2 000 000 mults)");
+                int n = _line.size();
+                for(int i=0; i< 2000000;i++){
+                    n *=i;
+                }
+                _packet.AddCertitude(n & 1);
```
I think I wanted to simulate a high CPU load to avoid optimization by the CPU (to test worst case scenarios without a specific filter), I can switch to a sleep

ThibaudCartegnie commented on samples/ftest/TestTask.cpp:
```diff
@@ -51,29 +52,37 @@ void TestTask::operator()() {
-                _certitudes.push_back(0);
+                DARWIN_ALERT_MANAGER.Alert(_line, 100, _packet.Evt_idToString(), "{\"filter_name\":\"" + darwin::Core::instance().GetName() + "\"}");
+                _packet.AddCertitude(0);
+                _response_body.append(_packet.GetBody());
```
To test that a response packet can be parsed with a body and certitudes if I remember correctly

frikilax replied:
ok

ThibaudCartegnie committed:
Fixed github TBE comments
- errors in comments
- missing parts of tests (and logging missing)
- removed useless variable in NextFilterConnector

frikilax commented:
Yeah, I'll try being lighter next time when suggesting changes: applying/validating them is a pain on github :sweat_smile:

frikilax commented:
Outside my remaining comments (which aren't really important, except the one on RedisManager.cpp), everything seems great!

@HugoSoszynski thoughts on the PR and remaining comments I made (especially [this one](https://github.com/VultureProject/darwin/pull/224#discussion_r775448936))?

ThibaudCartegnie committed:
Fixed last few issues with comments

ThibaudCartegnie committed:
Adapted dga for new usage of darwin api

ThibaudCartegnie committed:
Adapted dga for darwin api (forgot file)

ThibaudCartegnie commented on toolkit/RedisManager.cpp:
```diff
@@ -171,18 +171,30 @@ namespace darwin {
+                threadData->_retryAttempts = 1; // this is the first attempt
                 /*  try first to connect to current active connection (potentially updated),
                     then search master in case of failure */
-                if(this->Connect() and this->IsMaster())
+                if ((this->Connect() && this->IsMaster()) || this->FindAndConnect() || this->FindAndConnectWithRateLimiting()) {
+                    threadData->_retryAttempts = 0;
                     return true;
-                else
-                    return this->FindAndConnect();
+                } else {
+                    return false;
+                }
+            }
+            else if(threadData->_retryAttempts < this ->_maxRetryAttempts) {
+                // less than 8 seconds since last discovery, retrying
+                threadData->_retryAttempts++;
+                if ((this->Connect() && this->IsMaster()) || this->FindAndConnect() || this->FindAndConnectWithRateLimiting()) {
+                    threadData->_retryAttempts = 0;
+                    return true;
+                } else {
+                    return false;
+                }
```
The calls to Connect() and FindAndConnect() have timeouts that ensure that calls can't be fired in a very limited time, I'm not sure adding a timer could be very useful.
If you really think this is necessary, I can add the following code in the else if (retry) part : 
`while(now < threadData->_lastDiscovery + this->_healthCheckDelay){
                    std::this_thread::yield();
                }`

frikilax replied:
After talking about it, that looks OK to me

ThibaudCartegnie committed:
MANAGER:: Removed todo

Removed TODO : No handling of tcp case needed in that part, nothing to be changed
